### PR TITLE
fix(Explore): Show the tooltip only when label does not fit the container in METRICS/FILTERS/GROUP BY/SORT BY of the DATA panel

### PR DESCRIPTION
**Pull request by geido**
### SUMMARY
This PR makes the tooltip appear only when the label does not fit the container for the METRICS/FILTERS/GROUP BY (with D&D enabled)/SORT BY of the DATA panel. It also consolidates the look and behavior of the Columns and Metrics across the 3 different implementations, including the one in Superset UI. Finally, this PR hides the tooltip when the drag event starts.

###  REQUIRES SUPERSET-UI PR MERGED 
Attention, this PR requires the related PR https://github.com/apache-superset/superset-ui/pull/1271 in Superset-UI to be merged first.

### REFACTOR REQUIRED

Refactoring is suggested to remove the 3 different implementations and consolidate the code. This PR only intention is to resolve the issue #15996 within the limitations of the current implementation.

### BEFORE

https://user-images.githubusercontent.com/60598000/128191895-db410832-df25-4727-8df5-4579e0cb39a9.mp4

https://user-images.githubusercontent.com/60598000/128191922-1bae0671-33ab-4432-8c19-6ab6758cb03d.mp4

### AFTER

https://user-images.githubusercontent.com/60598000/128192168-606afa75-6634-48b9-b2e7-0e1036a2af5d.mp4

https://user-images.githubusercontent.com/60598000/128192223-080a903c-a103-4ae8-8a29-4273f4d8da88.mp4

### TESTING INSTRUCTIONS
1. Visit Explore
2. Add Columns and Metrics in Filters and Sort options either with drag and drop enabled or not
3. Observe the labels with long text and make sure they are truncated with ellipsis
4. Hover the labels and check the tooltip
5. Expand the panel until the text isn't truncated and hover the labels
6. Make sure the tooltip isn't appearing
7. Drag an item
8. Make sure the tooltip isn't appearing

### ADDITIONAL INFORMATION

- [x] Has associated issue: #15996
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by geido** [comment]
@rosemarie-chiu @jinghua-qa

**Comment by junlincc** [comment]
Insane!! you made it work!! 

please spin up the testenv with DND enabled 🙏

**Comment by geido** [comment]
For clarity:

- This PR only affects the METRICS/FILTERS/GROUP BY (with D&D enabled)/SORT BY of the DATA panel and nothing else. It does not change the behavior of the Metrics and Columns anywhere else, including the ones in the Dataset panel, for which an additional implementation will be required.

- Specifically for the GROUP BY when D&D is off, the component is using the _about-to-be-deprecated_ react-select, so it does not make sense to invest in it as it is not a blocker for the original issue and should be resolved when the Antdesign Select component will replace it.

Ideally, we should strive to implement the ability to show the tooltip only when it does not fit the container everywhere. However, this is an effort that will require several iterations.

**Comment by geido** [comment]
This PR needs to wait for https://github.com/apache/superset/pull/15942 to be merged as that PR contains the required Superset-UI changes. The CI checks won't pass until that's merged in.

**Comment by villebro** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_UX_BETA=true

**Review by villebro — APPROVED** [review]
LGTM - this is a really nice functional improvement and code also looks much cleaner now 👍

File: superset-frontend/src/explore/components/controls/DndColumnSelectControl/OptionWrapper.tsx [hunk 1]
```diff
@@ -28,8 +28,21 @@ import {
...
 import Option from './Option';
 
+export const OptionLabel = styled.div`
+  width: 100%;
+  overflow: hidden;
+  text-overflow: ellipsis;
+  white-space: nowrap;
+`;
+
+const LabelText = styled.span``;
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/DndColumnSelectControl/OptionWrapper.tsx:44]: Did you forget to add the props? If not, then it's better to remove this and use `span` directly.

File: superset-frontend/src/explore/components/controls/OptionControls/index.tsx [hunk 2]
```diff
@@ -195,7 +199,9 @@ export const OptionControlLabel = ({
 }) => {
   const theme = useTheme();
   const ref = useRef<HTMLDivElement>(null);
-  const [, drop] = useDrop({
+  const labelRef = useRef<HTMLDivElement>(null);
+  const hasMetricName = savedMetric?.metric_name;
+  const [_, drop] = useDrop({
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/OptionControls/index.tsx:204]:
```suggestion
  const [, drop] = useDrop({
```

File: superset-frontend/src/explore/components/controls/DndColumnSelectControl/OptionWrapper.tsx (hunk 1 above)
Comment (geido) [reply superset-frontend/src/explore/components/controls/DndColumnSelectControl/OptionWrapper.tsx:44]: I think this was meant to get some custom style and then it didn't require it. I'll remove it.

**Comment by michael-s-molina** [comment]
@geido I was manually testing and you can check in your after video (dnd enabled) that the Filters continue with ellipsis after expanding the panel. The Metrics and Group by are working, but the Filters are not.

**Review by michael-s-molina — APPROVED** [review]

**Comment by jinghua-qa** [comment]
@geido why do we need 2 tooltip for same metrics here?
Screen Shot 2021-08-05 at 10 50 50 AM

**Comment by michael-s-molina** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_UX_BETA=true

**Comment by geido** [comment]
> @geido why do we need 2 tooltip for same metrics here?
> Screen Shot 2021-08-05 at 10 50 50 AM
> 

@jinghua-qa it appears that this can be reproduced on Safari only. However, this is not related to this PR and we would need to address it separetely on Safari on a general level.

**Comment by junlincc** [comment]
@geido thanks for the fix! We also need to make the change of the data panel. tooltip should only show when it doesn't fit in the datapanel 🙏
Screen Shot 2021-08-05 at 10 14 21 AM

**Comment by junlincc** [comment]
> @jinghua-qa it appears that this can be reproduced on Safari only. However, this is not related to this PR and we would need to address it separetely on Safari on a general level.

Yes, i can confirm it;s not related to this PR. let's have a follow up one to clean up in Safari

**Comment by junlincc** [comment]
Also in filter, when there are multiple value selected, should we? we should? show all the selected values in the tooltip, instead of "..." this change can be made in this PR, maybe? @geido 

https://user-images.githubusercontent.com/67837651/128416984-6647d1ff-977d-4e3d-979a-dcb067aaf96e.mov

**Review by junlincc — CHANGES_REQUESTED** [review]
All selected filter values show show in the tooltip 🙏

**Comment by villebro** [comment]
> We also need to make the change of the data panel. tooltip should only show when it doesn't fit in the datapanel 🙏
> Screen Shot 2021-08-05 at 10 14 21 AM

Agreed, this is a major pain! But I vote for catching that in a follow-up.

**Comment by geido** [comment]
##### Tooltip in Filter IN - NOT IN

I have just updated the PR to show the tooltip whenever the tooltip title is different from the label.

In that way, the tooltip will show up in cases such as the "Filters" IN and NOT IN and in any other case where we might want to force-show the tooltip independently from the label content.

https://user-images.githubusercontent.com/60598000/128489310-08a019b5-2ad3-4673-851a-6a29285b637c.mp4

##### Tooltips in Data panel

This requires a separate implementation and it might require more changes than this one from a first look. @junlincc let me know what kind of priority we want to give it. I'll handle it in a separate PR.

##### Tooltips on Safari

As for the issue on Safari, I'd solve it in a separate PR because I want to apply a general solution and that will require a great amount of testing. I am working on it now.

CC @junlincc @villebro @michael-s-molina @jinghua-qa

**Comment by geido** [comment]
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true FEATURE_DASHBOARD_CROSS_FILTERS=true FEATURE_ENABLE_EXPLORE_DRAG_AND_DROP=true FEATURE_UX_BETA=true

**Review by michael-s-molina — APPROVED** [review]

**Review by junlincc — APPROVED** [review]
Product sign-off!! tested locally. will test again when both this and https://github.com/apache/superset/pull/16192 go in.

File: superset-frontend/src/explore/components/controls/DndColumnSelectControl/OptionWrapper.tsx [hunk 3]
```diff
@@ -107,6 +121,51 @@ export default function OptionWrapper(
     },
   });
 
+  const shouldShowTooltip =
+    (!isDragging && tooltipTitle && label && tooltipTitle !== label) ||
+    (!isDragging &&
+      labelRef &&
+      labelRef.current &&
+      labelRef.current.scrollWidth > labelRef.current.clientWidth);
```
Comment (rusackas) [inline superset-frontend/src/explore/components/controls/DndColumnSelectControl/OptionWrapper.tsx:129]:
```suggestion
      labelRef?.current?.scrollWidth > labelRef.current.clientWidth);
```
If you're feelin' fancy ¯\\\_(ツ)_/¯

**Pull request merged**